content/articles/how-to-build-a-web-crawler-with-java/index.md

```diff
@@ -0,0 +1,178 @@
+
+![breadth-first search](/bfs.png)
+
+In the diagram above, the traversal begins from Node 1, the parent or source node. Then it proceeds to its children - Nodes 2, 3, 4 in the second layer. From Node 4, the traversal continues to the leftmost part of the third layer and travels horizontally to the end of the layer. This continues until all nodes are visited.
+
+BFS has a time complexity of O(V+E), where V denotes the number of vertices and E, the number of edges. 
+
+You can learn more about BFS [in this tutorial](https://www.section.io/engineering-education/breadth-first-search/) that implements BFS to solve a maze.
+
+
+### Building the Web crawler
+
+Now to the tutorial’s core, you’ll build a web crawler that uses the BFS algorithm to traverse web pages. The crawler will begin from a source URL then visit every URL contained. Once every URL in this source URL has been visited, the algorithm will then visit each of the URLs in the children URLs and down the chain until it reaches a breakpoint that you will specify. The breakpoint here will represent how many URLs you want your web crawler to visit so that it doesn’t continue endlessly. 
+
+The algorithm will visit only URLs that have not been previously visited to ensure we don’t go in cycles. These URLs represent the vertices, and the connections between the URLs are the edges.
+
+#### The Pseudocode
+
+Start by adding the root URL to a queue and a list of visited URLs.
+While the queue is not empty, remove the URL from the queue and read its raw HTML content.
+While reading the HTML content of the URL, search for any other URL contained in the parent HTML.
+When a new URL is found, verify that it has not been previously visited by checking the list of visited URLs.
+Add the newly found unvisited URL to the queue and the list of visited URLs.
+Repeat steps 4 and 5 for every new URL found within the HTML content.
+When all the URLs in the HTML have been found, repeat from step two until the program reaches your specified breakpoint.
+
+####  Implementation
+
+Create a Java class with the name `WebCrawler and add the following code to the file:
+
+```java
+
+
+
+public class WebCrawler {
+
+    private Queue<String> urlQueue;
+    private List<String> visitedURLs;
+
+    public WebCrawler() {
+        urlQueue = new LinkedList<>();
+        visitedURLs = new ArrayList<>();
+    }
+
+    public void crawl(String rootURL, int breakpoint){
+        urlQueue.add(rootURL);
+        visitedURLs.add(rootURL);
+
+        while(!urlQueue.isEmpty()){
+
+            // remove the next url string from the queue to begin traverse.
+            String s = urlQueue.remove();
+            String rawHTML = "";
+            try{
+                // create url with the string.
```

Comment:
Can we bring down this bigger code snippet into smaller code snippets?

I guess making this code modular would allow you to do so.

Reply:
Break down the code snippets and explain the codes using bulleted points.

Reply:
I don't think the code is big enough for me to modularise. I have however split it into two methods and added comments.

Reply:
I expected that the comments I added to the code snippets would be sufficient to give the reader an idea of what is happening in the code before they get to the explanation section.